src/devices/cpu/upd177x/upd177x.cpp

```diff
@@ -0,0 +1,794 @@
+	state_add(UPD177X_SKIP_SAVE, "SKIP'", m_skip_save);
+
+	save_item(NAME(m_pc));
+	save_item(NAME(m_sp));
+	save_item(NAME(m_a));
+	save_item(NAME(m_a_save));
+	save_item(NAME(m_h));
+	save_item(NAME(m_n));
+	save_item(NAME(m_timer));
+	save_item(NAME(m_x));
+	save_item(NAME(m_y));
+	save_item(NAME(m_pa));
+	save_item(NAME(m_pb));
+	save_item(NAME(m_ram));
+	save_item(NAME(m_md0));
+	save_item(NAME(m_md1));
+	save_item(NAME(m_pnc1));
+	save_item(NAME(m_pnc2));
+	save_item(NAME(m_skip));
+	save_item(NAME(m_skip_save));
+	save_item(NAME(m_ts));
+	save_item(NAME(m_ns));
+	save_item(NAME(m_ss));
+	save_item(NAME(m_dac));
+	save_item(NAME(m_dac_sign));
+	save_item(NAME(m_tn_int));
+	save_item(NAME(m_counter));
+
+	set_icountptr(m_icount);
+
+	m_channel = stream_alloc(0, 1, clock() / 4);
+}
+
+
+void upd177x_cpu_device::device_reset()
+{
+	m_pc = 0;
+	m_sp = 0;
+	m_a = 0;
+	m_a_save = 0;
+	m_h = 0;
+	m_n = 0;
+	m_x = 0;
+	m_y = 0;
+	m_pa = 0;
+	m_pb = 0;
+	m_md0 = 0;
+	m_md1 = 0;
+	m_skip = false;
+	m_skip_save = false;
+	m_ts = false;
+	m_ns = false;
+	m_ss = false;
+	m_dac = 0;
+	m_dac_sign = true;
+	m_tn_int = false;
+	m_counter = 0;
+	m_timer = 0;
+	set_noise_counter_bit();
+}
+
+
+std::unique_ptr<util::disasm_interface> upd177x_cpu_device::create_disassembler()
+{
+	return std::make_unique<upd177x_disassembler>();
+}
+
+
+void upd177x_cpu_device::op0xxx(u16 opcode)
+{
+	switch (opcode)
+	{
+	case 0x0000:  // NOP
+		break;
+	case 0x0002:  // OUT PA
+		m_pa = m_a;
+		break;
+	case 0x0004:  // OUT PB
+		m_pb = m_a;
+		m_pb_out_cb(m_pb);
+		break;
+	case 0x0008:  // MOV X,RG
+		m_x = m_pnc1 & 0x7f;
+		break;
+	case 0x0201:  // MOV N,A
+		m_n = m_a;
+		break;
+	case 0x0208:  // MOV X,A
+		m_x = m_a & 0x7f;
+		break;
+	case 0x0401:  // IN PA
+		m_a = m_pa;
+		break;
+	case 0x0402:  // IN PB
+		m_a = m_pb;
+		break;
+	case 0x0404:  // RAR
+		m_a = ((m_a & 0x01) << 7) | (m_a >> 1);
+		break;
+	case 0x0408:  // RAL
+		m_a = ((m_a & 0x80) >> 7) | (m_a << 1);
+		break;
+	case 0x0502:  // OUT DA
+		m_channel->update();
+		m_dac = (m_ss ^ m_ts) ? ~m_a : m_a;
+		m_dac_sign = m_ss;
+		break;
+	case 0x0504:  // MUL1
+		if (!BIT(m_y, 0))
+		{
+			m_a = m_a >> 1;
+		}
+		m_y = m_y >> 1;
+		break;
+	case 0x050c:  // MUL2, usually gets executed 5 times to calculate X * (Y / 32)
+		if (BIT(m_y, 0))
+		{
+			m_a = (m_a + m_x) >> 1;
+		}
+		else
+		{
+			m_a = m_a >> 1;
+		}
+		m_y = m_y >> 1;
+		break;
+	case 0x0800:  // RET
+		m_sp = (m_sp - 1) & 0x07;
+		m_pc = (m_ram[STACK_START + (m_sp * 2) + 1] << 8) | m_ram[STACK_START + (m_sp * 2)];
+		break;
+	case 0x090f:  // RETI
+		m_sp = (m_sp - 1) & 0x07;
+		m_pc = (m_ram[STACK_START + (m_sp * 2) + 1] << 8) | m_ram[STACK_START + (m_sp * 2)];
+		m_a = m_a_save;
+		m_skip = m_skip_save;
+		m_tn_int = false;
+		break;
+
+	case 0x0005:  // STF
+	case 0x0101:  // MON
+	case 0x0501:  // JMPA
+	case 0x0602:  // OFF
+	case 0x0801:  // RETS
+	default:
+		fatalerror("%04x: Unsupported instruction %04x\n", m_pc - 1, opcode);
```

Comment:
You should at least include the device tag in the fatal error messages – it isn’t included automatically like it is for `logerror`.

Reply:
I have updated all fatalerrors to include the tag.